If the user has "fixendofline" set, append a blank line to preserve the EOF newline

bryanforbes opened the pull request:
Fixes #301

prabirshrestha commented:
@greghensley Can you try this. This uses a bit different implementation compared to this PR https://github.com/prabirshrestha/vim-lsp/pull/267. But given that LSP doesn't care about line ending I think this would work too.

prabirshrestha commented on autoload/lsp.vim:
```diff
@@ -235,6 +235,14 @@ function! s:on_text_document_did_close() abort
     call lsp#log('s:on_text_document_did_close()', l:buf)
 endfunction
 
+function! s:get_buffer_lines(buf) abort
```
Can we move this to util and make it public but start with `_` and have tests for it. We have regressed several times on this so having tests would help here. `lsp#utils#buffer#_get_lines(buf)`

bryanforbes replied:
Yep! I'll make the change later tonight.

bryanforbes commented:
> Could you add tests for this.

Absolutely, but how do I run the tests? I don't see any information about it.

bryanforbes commented:
I just took a look at @greghensley's PR and I think it would be good to incorporate it into the utility function @prabirshrestha would like me to split out. If it's ok with @greghensley, I'll combine our efforts into this PR and write a test.

mattn commented on autoload/lsp.vim:
```diff
@@ -235,6 +235,14 @@ function! s:on_text_document_did_close() abort
+		let l:lines += ['']
+	endif
+	return l:lines
+endfunction
```
Please use 4 spaces instead of tab character.

prabirshrestha commented:
Should be ok to combine.

We do need to document how to run tests. In the mean time you can follow instruction in docs at 
https://github.com/thinca/vim-themis and run `themis --reporter spec`.

bryanforbes commented:
Ok, I've split the utility function out; improved the detection based on the Vim 7.3 (which didn't have `fixendofline`) and 8.0 documentation for `endofline`, `binary`, and (in the case of 8.0) `fixendofline`; and added tests for all of the cases.

prabirshrestha commented:
Merged. Thanks for the fix as well as tests. Test will help us to never regress this again 😄
